qiskit_experiments/data_processing/data_processor.py

```diff
@@ -125,79 +190,117 @@ def _call_internal(
+                        unp.std_devs(data),
+                        index,
+                    )
+                )
+
+        # Return only first entry if len(data) == 1, e.g. [[0, 1]] -> [0, 1]
+        if data.shape[0] == 1:
+            data = data[0]
 
         if with_history:
-            return datum_, error_, history
+            return data, history
         else:
-            return datum_, error_
+            return data
 
-    def train(self, data: List[Dict[str, Any]]):
+    def train(self, data: Union[Dict, List[Dict]]):
         """Train the nodes of the data processor.
 
         Args:
             data: The data to use to train the data processor.
         """
-
         for index, node in enumerate(self._nodes):
             if isinstance(node, TrainableDataAction):
                 if not node.is_trained:
                     # Process the data up to the untrained node.
-                    node.train(self._call_internal(data, call_up_to_node=index)[0])
+                    node.train(self._call_internal(data, call_up_to_node=index))
 
-    def _data_extraction(self, data: Union[Dict, List[Dict]]) -> List:
+    def _data_extraction(self, data: Union[Dict, List[Dict]]) -> np.ndarray:
         """Extracts the data on which to run the nodes.
 
         If the datum is a list of dicts then the data under self._input_key is extracted
-        from each dict and appended to a list which therefore contains all the data. If the
-        data processor has to_array set to True then the list will be converted to a numpy
-        array.
+        from each dict and appended to a list which therefore contains all the data.
 
         Args:
-            data: A list of such dicts where the data is contained under the key self._input_key.
+            data: A list of such dicts where the data is contained under the key
+                ``self._input_key``.
 
         Returns:
             The data formatted in such a way that it is ready to be processed by the nodes.
 
         Raises:
             DataProcessorError:
                 - If the input datum is not a list or a dict.
-                - If the data processor received a single datum but requires all the data to
-                  process it properly.
                 - If the input key of the data processor is not contained in the data.
+                - If the data processor receives multiple data with different measurement
+                  configuration, i.e. Jagged array.
         """
         if isinstance(data, dict):
             data = [data]
 
+        data_to_process = []
+        dims = None
+        for datum in data:
+            try:
+                outcome = datum[self._input_key]
+            except TypeError as error:
+                raise DataProcessorError(
+                    f"{self.__class__.__name__} only extracts data from "
+                    f"lists or dicts, received {type(data)}."
+                ) from error
```

Comment:
This check should come earlier.

Reply:
This indicates i-th element of the list is not dictionary, i.e. `datum[key]` resulted in type error. We can exclude this from the loop, but we need
```python
if not all(isinstance(datum, dict) for datum in data):
    raise DataProcessorError(...)
```
seems bit inefficient.

Reply:
Fair enough.